rust/arrow/src/datatypes.rs

```diff
@@ -49,55 +54,108 @@ use crate::error::{ArrowError, Result};
+/// [the physical memory layout of Apache Arrow](https://arrow.apache.org/docs/format/Columnar.html#physical-memory-layout).
 #[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
 pub enum DataType {
+    /// A boolean datatype representing the values `true` and `false`.
     Boolean,
+    /// A signed 8-bit integer.
     Int8,
+    /// A signed 16-bit integer.
     Int16,
+    /// A signed 32-bit integer.
     Int32,
+    /// A signed 64-bit integer.
     Int64,
+    /// An unsigned 8-bit integer.
     UInt8,
+    /// An unsigned 16-bit integer.
     UInt16,
+    /// An unsigned 32-bit integer.
     UInt32,
+    /// An unsigned 64-bit integer.
     UInt64,
+    /// A 16-bit floating point number.
     Float16,
+    /// A 32-bit floating point number.
     Float32,
+    /// A 64-bit floating point number.
     Float64,
-    /// A timestamp with an optional timezone
+    /// A timestamp with an optional timezone.
+    ///
+    /// Time is measured as a Unix epoch, counting the seconds from
+    /// 00:00:00.000 on 1 January 1970, excluding leap seconds,
+    /// as a 64-bit integer.
+    ///
+    /// The time zone is a string indicating the name of a time zone, one of:
+    ///
+    /// * As used in the Olson time zone database (the "tz database" or
+    ///   "tzdata"), such as "America/New_York"
+    /// * An absolute time zone offset of the form +XX:XX or -XX:XX, such as +07:30
     Timestamp(TimeUnit, Option<Arc<String>>),
+    /// A 32-bit date representing the elapsed time since UNIX epoch (1970-01-01)
+    /// in days (32 bits).
     Date32(DateUnit),
+    /// A 64-bit date representing the elapsed time since UNIX epoch (1970-01-01)
+    /// in milliseconds (64 bits).
     Date64(DateUnit),
+    /// A 32-bit time representing the elapsed time since midnight in the unit of `TimeUnit`.
     Time32(TimeUnit),
+    /// A 64-bit time representing the elapsed time since midnight in the unit of `TimeUnit`.
     Time64(TimeUnit),
+    /// Measure of elapsed time in either seconds, milliseconds, microseconds or nanoseconds.
     Duration(TimeUnit),
+    /// A "calendar" interval which models types that don't necessarily
+    /// have a precise duration without the context of a base timestamp (e.g.
+    /// days can differ in length during day light savings time transitions).
     Interval(IntervalUnit),
+    /// Opaque binary data of variable length.
     Binary,
+    /// Opaque binary data of fixed size.
+    /// Enum parameter specifies the number of bytes per value.
     FixedSizeBinary(i32),
+    /// A variable-length string in Unicode with UTF-8 encoding.
     Utf8,
+    /// A list of some logical data type with variable length.
     List(Box<DataType>),
+    /// A list of some logical data type with fixed length.
     FixedSizeList(Box<DataType>, i32),
+    /// A nested datatype that contains a number of sub-fields.
     Struct(Vec<Field>),
     Dictionary(Box<DataType>, Box<DataType>),
```

Comment:
Could you add something for `Dictionary` as well?

Reply:
Oh yes, absolutely! I left it out because I wasn't really sure about it and then I forgot about it :sweat_smile: 

For clarification: what exactly is this `Dictionary`?
Is this conceptually similar to https://github.com/apache/arrow/blob/master/format/Schema.fbs#L269?
Also die C++ doc on this wasn't really clear, but also mentions `Maps`: https://arrow.apache.org/docs/cpp/api/datatype.html#_CPPv4N5arrow4Type10DICTIONARYE